rfd/0010-api.md

```diff
@@ -0,0 +1,235 @@
+---
+authors: Alexander Klizhentas (sasha@gravitational.com), Brian Joerger (bjoerger@gravitational.com)
+state: discussion
+---
+
+# RFD 10 - API and Client Libraries
+
+## What
+
+- Document and publish API OpenAPI spec and documentation.
+- Unify Golang Client experience.
+
+## Why
+
+There are several problems with Go client libraries:
+
+* Go library example pulls entire teleport dependency including `lib/auth` and `lib/services`.
+* There is no separate client library with guarantees of compatibility with specific version of Teleport.
+See examples [here](https://github.com/kubernetes/client-go#compatibility-matrix).
+*  It's impossible to generate library in any other language other than Go due to unclear API surface.
+*  Some client logic is residing in [plugins](https://github.com/gravitational/teleport-plugins/blob/master/access/access.go), while other is in `lib/auth`.
+* Code in `lib/auth` uses some concepts unfamiliar to Go users:
+
+```go
+// connectClient establishes a gRPC connection to an auth server.
+func connectClient() (*auth.Client, error) {
+    tlsConfig, err := LoadTLSConfig("certs/api-admin.crt", "certs/api-admin.key", "certs/api-admin.cas")
+    if err != nil {
+        return nil, fmt.Errorf("Failed to setup TLS config: %v", err)
+    }
+
+    // must prase addr list panics and is not necessary
+    authServerAddr := utils.MustParseAddrList("127.0.0.1:3025")
+    clientConfig := auth.ClientConfig{Addrs: authServerAddr, TLS: tlsConfig}
+
+    // TLS client is the only client
+    return auth.NewTLSClient(clientConfig)
+```
+
+## Details
+
+### Approach
+
+Keep Teleport's API layer unified, avoid having internal API vs external API layer.
+The access to the API surface is limited by role based access control.
+
+Avoid large refactors, instead use Go's type aliases to move things around while keeping things backwards compatible.
+
+```go
+// Preserve full backwards compatibility
+type services.Role = types.Role
+type auth.Service = proto.Service
+```
+
+Use Swagger and OpenAPI to document and create SDKs.
+
+#### Move protobuf types and services
+
+Point protobuf generators to generate types to another folder and use a system of submodules:
+
+```bash
+$ ls github.com/gravitational/teleport/api
+# root folder with API specs and is a submodule
```

Comment:
git submodule?
WDYT about keeping everything in one repo but declaring separate Go modules within?
the checkout size for users will be larger, but it'll be easier for us to keep versions in sync.

Reply:
This is meant as Go submodule, I'll update it